### ISSUE TITLE
Define behavior of descriptor-typed fields on dataclasses

### ISSUE THREAD

**Issue by fff8dfff-07d0-462f-b0dc-e5943c532afd** [python/cpython#91330]
BPO | [47174](https://bugs.python.org/issue47174)
--- | :---
Nosy | @ericvsmith, @JelleZijlstra, @debonte

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2022-03-30.19:00:12.315>
labels = ['type-feature', 'library', '3.11']
title = 'Define behavior of descriptor-typed fields on dataclasses'
updated_at = <Date 2022-04-01.20:03:22.690>
user = 'https://github.com/debonte'
```

bugs.python.org fields:
```python
activity = <Date 2022-04-01.20:03:22.690>
actor = 'zzzeek'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2022-03-30.19:00:12.315>
creator = 'debonte'
dependencies = []
files = []
hgrepos = []
issue_num = 47174
keywords = []
message_count = 1.0
messages = ['416392']
nosy_count = 4.0
nosy_names = ['eric.smith', 'zzzeek', 'JelleZijlstra', 'debonte']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue47174'
versions = ['Python 3.11']
```

</p></details>

**Comment by ericvsmith** [comment]
I think it's fine to document and test for the first two items.

1. The value passed to __init__ for that field is passed to the descriptor’s __set__ method, rather than overwriting the descriptor object.
2. Getting/setting the value of that field goes through __get__/set, rather than getting/overwriting the descriptor object.

For:
3. If a field is annotated with a descriptor type but is *not* assigned a descriptor object as its default value, it acts like a non-descriptor field.

I think that's the correct behavior. To do otherwise would require dataclasses to look at the annotation to figure out what type of descriptor to create, which I think is outside its scope.

For:
4. When a field with a descriptor object as its default value is initialized (when the code for the dataclass is initially executed), __get__ is called with a None instance and the return value is used as the field's default value. See the example below.

I don't really understand what's going on. I'll look at it further.

**Comment by carljm** [comment]
Behavior 4 happens because in `_get_field` dataclasses does `default = getattr(cls, a_name, MISSING)` to get the default value (if any) off the class (where it naturally is found since default values are syntactically defined as class attribute values.) If `cls.a_name` has been set to a descriptor, the normal behavior of a descriptor when it is accessed directly off the class is to call its `__get__` method with the obj/instance argument set to `None`.

I actually think this behavior is quite nice in practice for dataclasses, and if we are going to document these behaviors, IMO it makes sense to document it too. It is nice because it gives the `__get__` method a consistent return type which is the type of both the default value and any other value returned from accessing the field on an instance, and gives descriptor fields a simple way (that already naturally falls out of the existing implementation) to specify their default value.

I think the possible downside is that it is somewhat common today for descriptors to `return self` if the `obj` argument is `None` -- i.e. if you access the descriptor on the class you get the actual descriptor object back. So there may be existing cases in the wild where people are using descriptors with this `return self` behavior on dataclass-like classes, and would have to change it in order to work with the current behavior of dataclasses. (SQLAlchemy may fall into that category?) On the other hand, as far as I know there often isn't a strong use case for the `return self` behavior -- it's just that there's no other obvious thing to return when accessed on a class and so `return self` is kind of a simple fallback that supports more transparent introspection of the class. So it may not be a major cost to move away from `return self`?

**Comment by debonte** [comment]
> I actually think this behavior is quite nice in practice for dataclasses, and if we are going to document these behaviors, IMO it makes sense to document it too.

@carljm, thanks for explaining why this works. I agree that it's worth documenting.

> I think the possible downside is that it is somewhat common today for descriptors to return self if the obj argument is None

Are you saying that this is a downside to documenting it? Given that dataclass already behaves this way and no one has expressed a desire to change that behavior, is there a downside?

I think your point is that if a user relies on the default value of a descriptor-typed field and the descriptor's `__get__` returns `self` for the `None` `obj` case, it's likely that the user will run into problems at runtime because they'll be passing a descriptor instance (the `self` instance) to the descriptor's `__set__` method. Is that correct?

So perhaps type checkers and language servers will choose not to support this because they don't want to lead users into that sort of problem. But documenting the behavior at least makes it common knowledge and can steer descriptor authors in the right direction in the future. And maybe tools can help there, for example maybe by warning if a `__get__` overload for `None` `obj` returns a descriptor type.

Do you disagree?

**Comment by zzzeek** [comment]
SQLAlchemy's descriptor (the one in question) does not return "self" when "obj" is None, it returns a SQL expression construct.  this doesn't actually affect SQLAlchemy with dataclasses in any case as we don't make any dataclasses with said descriptor interpreted in a dataclasses context at runtime.  

as far as descriptors generally reutnring "self" when obj is None, I'm not sure what else you think would be returned by the average "normal" descriptor.  returning "self" is usually kind of important for tools like Sphinx autodoc and such to work, and also the [primary document on descriptors](https://docs.python.org/3/howto/descriptor.html#properties) pretty much uses "return self" in all the examples.

**Comment by debonte** [comment]
@zzzeek, are you arguing for not documenting this behavior at all or just pointing out that this is only relevant to descriptors that are used as dataclass field types rather than all descriptors?

**Comment by zzzeek** [comment]
I am sort of pointing out the latter, but also wondering what the suggestion is for garden variety descriptors to do when their `__get__()` method is called with `obj=None`.

**Comment by debonte** [comment]
I think that may be intentionally left open-ended. [PEP 252](https://peps.python.org/pep-0252/#specification-of-the-attribute-descriptor-api) says:
> `__get__()`: ...This is also referred to as a “binding” operation...Exactly what is returned by the binding operation depends on the semantics of the descriptor

**Comment by carljm** [comment]
> Are you saying that this is a downside to documenting it? Given that dataclass already behaves this way and no one has expressed a desire to change that behavior, is there a downside?

No, I don't really think there is a downside to documenting it. I guess I meant "downside" broadly as in "this behavior kind of conflicts with a typical way that descriptors are written," but I don't think that's sufficient reason either to change dataclasses' behavior or avoid documenting it. I'm not sure how common use of descriptors as dataclass fields will ever be, but I think if they are used, this way of setting the default value makes as much sense as anything, and has some strong advantages (not least that it is the status quo).

It may also be worth documenting the corollary that ` if obj is None: raise AttributeError()` is how a descriptor `__get__()` method would signal to dataclasses that there is no default value for the field.

@zzzeek I wasn't trying to question `return self` as one reasonable choice of `obj=None` behavior for descriptors that aren't meant for use as dataclass fields.

**Comment by debonte** [comment]
@ericvsmith, [my PR](https://github.com/python/cpython/pull/94424) got merged without your review. Are you OK with the test changes in there?

**Comment by ambv** [comment]
Yes, this is now landed in 3.10 - 3.12. Thanks, Erik!

I'll leave the issue open for a while in case Eric's got any test improvement suggestions. We can easily make those improvements forward.

**Comment by ericvsmith** [comment]
Thanks, @debonte. I thought I'd reviewed this but I guess I never committed it. It looks good to me, so I'm closing this.
